### ISSUE TITLE
Move controller bulbs turn off after a while

### ISSUE THREAD

**Issue by connersc** [cboulay/PSMoveService#246]
Hi all, hopefully someone can help with this issue.

I'm using FreePIE etc. for head tracking, Riftcat to connect and I get this issue when using SteamVR. I haven't really done much testing outside of SteamVR. I've got a PS Move controller mounted to my HMD (cardboard type) and 2x PS Move controllers in my hands. I'm using the latest release from 9 days ago.

It's a bit jittery but working well, I'm sure it will only get better, thanks to HipsterSloth, CBoulay and everyone else working on the development.

After playing for a minute or so controller 1's bulb will turn off and the controller will become unresponsive. Give it another minute or so and the other 2 will turn off too. The controllers still have the red light to indicate that they're on, but seem unresponsive. I don't have this issue when using the Config tool for 10-15 mins at a time.

**Comment by HipsterSloth** [comment]
Someone else had reported a similar issue a while back but I could never reproduce it. It's possible there is something particular going on with FreePie also connected to PSMoveService. Would you mind posting your `Program Files (x86)\Steam\logs\vrserver.txt` log and the log from the PSMoveService window? Hopefully there is something in there indicating why the controller is turning off.

**Comment by connersc** [comment]
Here's the logs. When I CTRL+C'd the PSMoveService window the service crashed (unresponsive, windows tried to kill).

I can quite predictably recreate this issue on my setup. If I start AHK script to start all the programs, then the SteamVR tutorial, by the time I get to the point where the Wheatley-a-like is telling you to press grip buttons, that's when the HMD and 2nd hand controller turn off.

[PS Move Log.txt](https://github.com/cboulay/PSMoveService/files/531790/PS.Move.Log.txt)
[vrserver.txt](https://github.com/cboulay/PSMoveService/files/531791/vrserver.txt)

**Comment by HipsterSloth** [comment]
Looks like your controllers are not sending data after a while: 

> ServerDeviceView::poll - Device id 1 closing due to no data (100 failed poll attempts)

I don't know if this is an issue with you controllers or psmoveservice.

Can you run test_console_client with only psmoveservice running and see if the same issue reproduces? It should connect to the first controller and just spam controller data to the console. 

How old are your controllers btw? Some older psmove controller can have issues holding a charge. You might also try turning the controller on, letting the battery drain completely, then let it charge overnight.

**Comment by connersc** [comment]
Thanks - I've had test_console_client going for nearly 40 mins now, no sign of the connection dropping, bulb's still on.
[test_console.txt](https://github.com/cboulay/PSMoveService/files/532112/test_console.txt)

I don't know how old the controllers are, I bought them all second hand. When the bulbs turn off though, the red light at the bottom of the controller is still on. 

Is it something to do with vive touchpad emulation? They seem to go quite predictably at those cues in the SteamVR tutorial.

Could it possibly have something to do with the bluetooth adapter? Maybe some power saving mode built in? I've got an unbranded bluetooth 3.0 adapter that I've had and used for ages. Also it's plugged into a USB 3.0 port on a PCIe host. I'm going to try it tomorrow in a motherboard 2.0 port, then I'll try it on it's own root hub. I've had to play with that sort of thing when I'm using 3-4 PS Eyes.

**Comment by HipsterSloth** [comment]
> Thanks - I've had test_console_client going for nearly 40 mins now, no sign of the connection dropping, bulb's still on.

Hmm that's interesting. Sounds like it's not a battery issue then. I would had said it was USB bandwidth issue with the cameras running at the same time, but test_console_client also opens the cameras as well (since it asks for the position data). Maybe there is some issue when there are two different clients connected to psmove service is connected as the same time (steamvr is one client, and FreePIE is another).

> Is it something to do with vive touchpad emulation? They seem to go quite predictably at those cues in the SteamVR tutorial.

I was working on a new 5.4 release today. I'll run through the steamvr setup tutorial myself and see if I can repro the issue at that same point in the tutorial.

> Could it possibly have something to do with the bluetooth adapter? Maybe some power saving mode built in? I've got an unbranded bluetooth 3.0 adapter that I've had and used for ages. Also it's plugged into a USB 3.0 port on a PCIe host. I'm going to try it tomorrow in a motherboard 2.0 port, then I'll try it on it's own root hub. I've had to play with that sort of thing when I'm using 3-4 PS Eyes.

Maybe? Usually the Bluetooth adapter failures I see have to do with the refusing to pair with the psmove controller in the first place. But the windows bluetooth stack can be so flakey I couldn't discount the possibility. Switching around USB ports is also a good idea just to eliminate a possibility.

I use the the ASUS BT-400 adapter: https://www.amazon.com/ASUS-USB-Adapter-Bluetooth-USB-BT400/dp/B00DJ83070, but I don't want to tell you to go out and buy another BT adapter just yet.

On an sort of related note I just noticed this on the Amazon page for the ASUS BT-400:

![boughttogether](https://cloud.githubusercontent.com/assets/16793853/19421381/790779f8-93b4-11e6-996f-53dd244cb6e4.jpg)

This warms my heart.

**Comment by connersc** [comment]
That's great! You're definitely having an impact on what we do :)

I've tried a brief test tonight with the BT adapter on it's own 2.0 root hub. I only tried it for 5 mins but the controllers all stayed on, I got to the point in the SteamVR tutorial where he's asking you to press the menu button, I haven't played with my mapping yet so that's as far as I can get. I think the magnetometer in one of my controllers is a bit squiffy, I'll probably earmark it as the HMD controller. Am I right in thinking that the HMD controller doesn't use the magnetometer, just the bulb?

I'll try an extended test this weekend with all 4 cameras running. Thanks again for the response, keep up the amazing work!

**Comment by HipsterSloth** [comment]
> I've tried a brief test tonight with the BT adapter on it's own 2.0 root hub. I only tried it for 5 mins but the controllers all stayed on, I got to the point in the SteamVR tutorial where he's asking you to press the menu button, I haven't played with my mapping yet so that's as far as I can get.

Interesting. Let me know if this continues to work with the adapter plugged into the new location. If it still continues to run fine after additional testing we can close this issue and I'll add some notes to the troubleshooting page reflecting this conversation in case anyone else runs across the same problem.

>  I think the magnetometer in one of my controllers is a bit squiffy, I'll probably earmark it as the HMD controller. Am I right in thinking that the HMD controller doesn't use the magnetometer, just the bulb?

I believe that is correct. I think Riftcat uses the IMU from your phone instead. As for the magnetometer in your 3rd psmove it might actually be ok. One of the problems with my current implementation of magnetometer calibration is that it's possible to not capture the true extents of your magnetometer sensor range. This can result in magnetometer readings that are slightly off resulting in drift. See notes from this issue:

https://github.com/cboulay/PSMoveService/issues/232

I want to fix this with a better magnetometer sample fitting algorithm at some point soon.

**Comment by Xierion** [comment]
I am having the same experience, however my issue only started when using Freepie to track all 3 gyro axis, as opposed to just yaw as per the example script.

My reasoning behind using a PSmove controller as opposed to the phone gyro is it appears the phone gyro drifts if frames drop from Riftcat/VRidge. Freepie tracking works perfectly, until all the controllers (HMD + 2 x hand controllers) drop off.

My PSMoveService console shows the following as the last lines:
[2016-10-23 17:17:25.239]: ClientConnection::handle_tcp_read_request_header - Failed to read header on connection 0: An existing connection was forcibly closed by the remote host
[2016-10-23 17:17:25.248]: ClientConnection::stop - Stopping client connection id 0

**Comment by Xierion** [comment]
On a side note, I have put my lounge room back to normal, which involved disconnecting 2 of the 4 PS3 Eye cameras, and the service/bulbs have not stopped in quite some time. Could this be caused by a USB bandwidth or processing power issue?

**Comment by HipsterSloth** [comment]
Oh interesting. It's possible that that an over-saturated USB root hub is preventing data coming from the Bluetooth adapter. You can tell which root hub a device is plugged into using USBView:

http://k0lee.com/dc3000/usbts.htm

**Comment by Xierion** [comment]
I've distributed the 4 cameras over different USB Root Hubs (2 each) and it appears to have done the trick. I did notice that the test_camera.exe program showed the 4 video feeds lagging when connected to the same hub, but smoothed when separated too.

**Comment by bclindner** [comment]
I had this problem but managed to fix it by connecting my Bluetooth adapter to another port, as the hub it was connected to was unreliable. Might help to try that, if nothing else works.

**Comment by Wully616** [comment]
I'm having the same issue, running using:
riftcat with a wireless 2.5GHz dongle.
3 ps motion controllers 
4 ps3 eye cameras
bluetooth adapter (not sure of brand)

All three controllers lights go off after 1 - 2 minutes in game.
At first I thought it was the freepie bridge, I disabled that, disconnected one of my cameras, tried moving the bluetooth adapter to another port, but no luck.

I'm starting to think it might be the 2.5GHz band of the wifi interfering with bluetooth on the controllers, causing the disconnects.

I've attached my log files. 
[vrserver.txt](https://github.com/cboulay/PSMoveService/files/679784/vrserver.txt)
[PSMoveService.txt](https://github.com/cboulay/PSMoveService/files/679785/PSMoveService.txt)
